Add Java module definitions

A248 opened the pull request:
Aims to solve #1012

I added Java module definitions where I thought appropriate. Several of the subprojects aren't designed for external consumption, such as test suites and benchmarks.

Packages named "internal" I treated as internal to chicory, i.e. not intended for external consumption.

Additionally, I did my best to check when types from a dependency were exposed in a module's public API, in order to use `requires transitive` appropriately. It's possible I missed a spot, but this will never be breaking and will result in at most a compiler warning (on chicory's side) or the need to write another `requires` (on the library user's side).

**Checkstyle**

Chicory was still using checkstyle 9.3, which is still the default version in the maven-checkstyle-plugin. Because this old checkstyle version was running into problems consuming `module-info.java`, I increased the version to the latest.

This required checking for breaking changes in checkstyle, which I did by looking at [every release since 9.3](https://checkstyle.org/releasenotes.html#Release_9.3). Checkstyle does not follow SemVer so you have to look at every release notes.

**Terminology**

For people reading this PR, the PR itself targets the Java Modules introduced in Java 9. They are a separate concept from WASM modules.

A248 commented:
Theoretically, yes. Some behavior is different on the module path: Java's `ServiceLoader` and `ClassLoader.getResource(AsStream)` are both affected by switching from an unnamed jar to a full module.

in practice, compilation will mostly ensure that the modules are declared properly. But if you want to spin up an integration test, you can always create a subproject that runs Chicory on the module path. This subproject can be created either as a plain Maven module, or using the maven-invoker-plugin.

Theoretically speaking though, there's no way to reach 100% coverage unless you literally fork all test runs and run them in a modular environment. That's why I rely on `rg ServiceLoader` and `rg getResource` when checking a codebase for the ability to use modules (using rg = ripgrep).

A248 commented on pom.xml:
```diff
@@ -554,6 +554,9 @@
           <excludeGeneratedSources>true</excludeGeneratedSources>
           <checkstyleRules>
             <module name="Checker">
+              <module name="BeforeExecutionExclusionFileFilter">
+                <property name="fileNamePattern" value="module\-info\.java$"/>
+              </module>
```
module-info is wholly unsupported by checkstyle, so this is the recommended configuration (per their docs) to skip it.

https://github.com/checkstyle/checkstyle/issues/8240

dmlloyd replied:
Makes sense. I'd put a comment in here to that effect though.

A248 commented:
Good news! Not only did the tests caught a couple bugs, but they showed that the "internal" packages need to be exported after al. This obviates the need for `-Xlint:-module` and makes the PR overall more clean.

All that considered, I don't have any other changes in mind to make to this PR, besides addressing your review concerns, of course.

andreaTP commented on compiler/pom.xml:
```diff
@@ -62,4 +62,23 @@
+            <configuration>
+              <compilerArgs>
+                <arg>--add-modules</arg>
+                <arg>java.sql</arg>
```
this is surprising

andreaTP commented on wasm-corpus/pom.xml:
```diff
@@ -19,6 +19,17 @@
 
   <build>
     <plugins>
+      <plugin>
```
thanks

andreaTP commented:
Just to spell it out, I was waiting an iteration from you @A248 on this PR, let me know if you are going to have time to progress on it :slightly_smiling_face: and thanks again :pray:
